src/xwindows/keyboard.rs

```diff
@@ -0,0 +1,590 @@
+                            xkb::x11::MIN_MINOR_XKB_VERSION
+                        );
+                    }
+                }
+                Err(_) => {
+                    panic!("could not check if xkb is supported");
+                }
+            }
+        }
+
+        let context = xkb::Context::new(xkb::CONTEXT_NO_FLAGS);
+        let device_id = xkb::x11::get_core_keyboard_device_id(&connection);
+        if device_id == -1 {
+            panic!("Couldn't find core keyboard device");
+        }
+
+        let keymap = xkb::x11::keymap_new_from_device(
+            &context,
+            &connection,
+            device_id,
+            xkb::KEYMAP_COMPILE_NO_FLAGS,
+        );
+        let state = xkb::x11::state_new_from_device(&keymap, &connection, device_id);
+
+        use std::ffi::CStr;
+        let locale = unsafe{CStr::from_ptr(libc::setlocale(libc::LC_CTYPE, std::ptr::null()))}
+            .to_str()
+            .expect("Failed to query locale");
+
+        let table =
+            xkb::compose::Table::new_from_locale(&context, locale, xkb::compose::COMPILE_NO_FLAGS)
+                .expect("Failed to get ctable");
+        let compose_state = xkb::compose::State::new(&table, xkb::compose::STATE_NO_FLAGS);
+
+        {
+            let map_parts = xcb::xkb::MAP_PART_KEY_TYPES
+                | xcb::xkb::MAP_PART_KEY_SYMS
+                | xcb::xkb::MAP_PART_MODIFIER_MAP
+                | xcb::xkb::MAP_PART_EXPLICIT_COMPONENTS
+                | xcb::xkb::MAP_PART_KEY_ACTIONS
+                | xcb::xkb::MAP_PART_KEY_BEHAVIORS
+                | xcb::xkb::MAP_PART_VIRTUAL_MODS
+                | xcb::xkb::MAP_PART_VIRTUAL_MOD_MAP;
+
+            let events = xcb::xkb::EVENT_TYPE_NEW_KEYBOARD_NOTIFY
+                | xcb::xkb::EVENT_TYPE_MAP_NOTIFY
+                | xcb::xkb::EVENT_TYPE_STATE_NOTIFY;
+
+            let cookie = xcb::xkb::select_events_checked(
+                &connection,
+                device_id as u16,
+                events as u16,
+                0,
+                events as u16,
+                map_parts as u16,
+                map_parts as u16,
+                None,
+            );
+
+            cookie
+                .request_check()
+                .expect("failed to select notify events from xcb xkb");
+        }
+
+        let kbd = Keyboard {
+            context: context,
+            device_id: device_id,
+            keymap: RefCell::new(keymap),
+            state: RefCell::new(state),
+            compose_state: RefCell::new(compose_state),
+        };
+
+        (kbd, first_ev, first_er)
+    }
+
+    pub fn process_key_event(
+        &self,
+        xcb_ev: &xcb::KeyPressEvent,
+    ) -> Option<(KeyCode, KeyModifiers)> {
+        let xcode = xcb_ev.detail() as xkb::Keycode;
+        let xsym = self.state.borrow().key_get_one_sym(xcode);
+        let pressed = (xcb_ev.response_type() & !0x80) == xcb::KEY_PRESS;
+        if pressed {
```

Comment:
I tend to prefer early-return over `if-then-else` as it results in fewer things for human brains to parse:

```
if !pressed {
   return None;
}
self.compose_state.borrow_mut().feed(xsym);
```